src/blas/backends/mklcpu/mklcpu_common.hpp

```diff
@@ -47,7 +47,9 @@ static inline auto host_task_internal(H &cgh, F f, int) -> decltype(cgh.run_on_h
 
 template <typename K, typename H, typename F>
 static inline void host_task_internal(H &cgh, F f, long) {
+#ifndef SYCL_DEVICE_ONLY
```

Comment:
Can you explain the reason for this change pls?

Reply:
As in hipSYCL `host_task` is not implemented we use `single_task` instead. Since `single_task` could be used as GPU kernel it is not allowed to call CPU functions such as the cblas functions. In order to prevent that from happening we exclude it from the compilation in case we are compiling for the device.

Reply:
Any reason why host_task is not implemented in hipSYCL?

Reply:
@mmeterel host tasks are generally inefficient for backend interop, which is however one of their primary use cases. Since hipSYCL provides more efficient alternatives, implementing host task is curently low on the priority list.

Reply:
> single_task is not allowed to call CPU functions?

HipSYCL compiles every kernel for all available platforms. For example, in the case, CPU and Cuda are targetted all kernels will be compiled for both. It is illegal to call host functions from device code; therefore we need to differentiate between when generating device and CPU code, which can be done using the `SYCL_DEVICE_ONLY` macro, so that we only compile the `single_task` kernels in case we are compiling for CPU. Since these functions are only meant to be executed on the CPU, it shouldn't be a problem that the GPU version does not exist.

> Also, did you mean to use SYCL_DEVICE_ONLY macro? Or are you proposing to add a new macro here?

`SYCL_DEVICE_ONLY` is set by hipSYCL in case generating device code. I think here this is the right macro to use. See [here](https://github.com/illuhad/hipSYCL/blob/develop/doc/macros.md)

Reply:
> I am not an expert in this area, but AFAIK, we use single_task to call CPU functions if run_on_host_intel is not supported by the compiler. So, this is confusing to me. Are you saying, in hipSYCL implementation, single_task is not allowed to call CPU functions?

As @sbalint98 points out, since `single_task()` is designed to execute kernel code, it will in general be also compiled for non-CPU devices. This is also the case with DPC++ and to be expected with any SYCL implementation.

> Also, did you mean to use `__SYCL_DEVICE_ONLY__` macro? Or are you proposing to add a new macro here?


hipSYCL currently cannot guarantee the existence of `__SYCL_DEVICE_ONLY__` on all backends. There is `SYCL_DEVICE_ONLY` instead, but you are right to point out that the SYCL spec says that it should be called `__SYCL_DEVICE_ONLY__`. This is an issue in hipSYCL that we should look at fixing.

For the sake of completeness, I'd like to point out that in general `__SYCL_DEVICE_ONLY__` is not sufficient to control whether code executes only on device, even though it is frequently misunderstood in that way (and this approach happens to work in DPC++ and most hipSYCL backends). `__SYCL_DEVICE_ONLY__` is defined by the spec to be set when the current compilation does not produce a host binary. This means that this macro only work for this purpose when the SYCL implementation relies on multiple independent compilation passes. If the SYCL implementation has a unified frontend compilation pass for host and device with late separation of device code at the IR level, this macro no longer works. In that case `__SYCL_SINGLE_SOURCE__` should be defined.

Reply:
@sbalint98  @illuhad Thanks for your patience with me :) and for your explanations. If I understand correctly, you will update this code to use __SYCL_DEVICE_ONLY__? Am I right?

Reply:
@mmeterel Yes, I think this should be the course of action. We have just merged support for `__SYCL_DEVICE_ONLY__` into hipSYCL for this purpose :-)

Reply:
Now using `__SYCL_DEVICE_ONLY` :)

Reply:
Thanks @sbalint98 Did you test if this breaks anything with DPC++ compiler?

Reply:
Yes I tested it with both dpcpp and hipSYCL and everything seemed to be fine

Reply:
Thanks for confirming.